app/src/main/cpp/droidvnc-ng.c

```diff
@@ -315,6 +315,17 @@ JNIEXPORT jboolean JNICALL Java_net_christianbeier_droidvnc_1ng_MainService_vncS
     theScreen->setXCutTextUTF8 = onCutText;
     theScreen->newClientHook = onClientConnected;
 
+    in_addr_t address = 0; // Default is 0.0.0.0
+    if (listenIf != NULL) {
+        const char *listenIfNative = (*env)->GetStringUTFChars(env, listenIf, NULL);
+        if (!rfbStringToAddr((char*)listenIfNative, &address)) {
+            address = 0; // If problems arise, assume 0.0.0.0
+        }
+        (*env)->ReleaseStringUTFChars(env, listenIf, listenIfNative);
+    }
+
+    // With the listenInterface property one can define where the server will be available
+    theScreen->listenInterface = address;
```

Comment:
I have to be honest, on this side I had no idea, but, looking at the source code for libvnc, it seems that internally, rfbServerInit uses rfbListenOnTCPPort for "auto-probing" at some point... we could use that to verify if the interface is up, I guess.

Reply:
Hi, I tried working on listening for Network events, but I'm having some problems.
I found a way to intercept Network State changes, but it still doesn't work as expected (for example, when I turn on/off the VPN it doesn't work), moreover, I don't know how to integrate it with the service, for now, only WLAN works for me...

The NetworkInterfaceTester is used, at the moment, only to update the spinner, but with some more work, we should be capable of using it to properly intercept network changes even in the service.

I think I need help though, I don't know how to integrate this "tester" with the Service (I still haven't learnt how to deal with those in Android).
Of course, even only suggestions would do, I'm not expecting others to do the work for me.

As for the .c file, if the address is not valid, an error is reported (instead of using 0.0.0.0).
I'm not sure how to deal with check if the interface is up in the c-side of things... the actual problem, is detecting the change in the used network interface.

Reply:
Maybe you don't have to handle the situation - if the server just keeps running and serves requests again after the interface is up again, that'd be OK. It should just not get into an unrecoverable state. I have not tested this at all, so just asking what you see :-)

Reply:
Oh, do not worry... actually, I was thinking of another approach on the c-side of things, except, well, I'm not well to informed with libvnc.
Is there a way to detect connections problems with it? Something like stopping rfbRunEventLoop and set some flag?

Reply:
As far as I can tell, the function vncIsActive always reports true, when listening to WLAN, even though Wi-Fi is removed.
Internally, vncIsActive uses rfbIsActive from libvnc which, it seems it does not take into account the listenInterface "attribute" of RfbScreen.

How about some kind of polling mechanism in order to see if the socket (listenSock in our case) is still "usable"?

Reply:
I can think of two possible ways to solve this:

1. Java-only: detect Network state change, if the server is listening to the interface that was removed, force-stop the server with vncServerStop;
2. C+Java: make it possible to automatically stop the server from C, by checking the connection state and stopping the rfb thread accordingly.
I guess network state detection may be used to update the UI only, since everything is done from C.

Reply:
I was able to implement a Java-only solution, it probably needs more polish though.
Also, for some reason, it didn't work with my VPN (mullvad), the callbacks were not called at all.
I tested with Wifi though, it should work with cellular data also.

Reply:
Problem with VPN was solved, it should work with any interface now.
I'm waiting for review.